### PR TITLE
KAFKA-12309 The revocation algorithm produces uneven distributions

### PR DESCRIPTION
**Pull request by chia7712**
issue: https://issues.apache.org/jira/browse/KAFKA-12309

# issue description
When adding a new worker, the revocation algorithm tries to revoke some connectors/tasks from existent workers so as to move them to new worker to balance load. However, the algorithm can revoke incorrect number of connectors/tasks for following cases.

### Assignments:
"worker_0" -> 8 connectors/tasks
"worker_1" -> 8 connectors/tasks
(New) "worker_2" -> 0 connectors/tasks

### Result
"worker_0" -> 6 connectors/tasks
"worker_1" -> 6 connectors/tasks
"worker_2" -> 4 connectors/tasks

### (expected) Result
"worker_0" -> 6 connectors/tasks
"worker_1" -> 5 connectors/tasks
"worker_2" -> 5 connectors/tasks
or 
"worker_0" -> 5 connectors/tasks
"worker_1" -> 6 connectors/tasks
"worker_2" -> 5 connectors/tasks

# root cause
the algorithm makes all existent workers keep `ceil` (`numToRevoke = existing.tasksSize() - ceilTasks;`) connectors/tasks. That results in that there is no enough revoked connectors/tasks to be assigned to new worker.

# extra changes included by this PR

1. move revocation algorithm to a helper method in order to test it

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/RebalanceSourceConnectorsIntegrationTest.java [hunk 1]
```diff
@@ -359,7 +359,9 @@ private boolean assertConnectorAndTasksAreUniqueAndBalanced() {
                     tasks.values().size(),
                     tasks.values().stream().distinct().collect(Collectors.toList()).size());
             assertTrue("Connectors are imbalanced: " + formatAssignment(connectors), maxConnectors - minConnectors < 2);
+            if (minConnectors > 1) assertEquals("Some workers have no connectors", connectors.size(), connect.workers().size());
```
Comment (chia7712) [inline connect/runtime/src/test/java/org/apache/kafka/connect/integration/RebalanceSourceConnectorsIntegrationTest.java:365]: make sure there is no idle worker

**Comment by chia7712** [comment]
@rhauch Could you take a look? this uneven distributions can be reproduced easily so it would be nice to fix it asap.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java [hunk 2]
```diff
@@ -562,50 +561,63 @@ private void resetDelay() {
...
+        int numberOfBiggers = numberOfActives % numberOfTotalWorks;
+        int ceil = numberOfBiggers == 0 ? floor : floor + 1;
         for (WorkerLoad existing : existingWorkers) {
-            Iterator<ConnectorTaskId> tasks = existing.tasks().iterator();
-            numToRevoke = existing.tasksSize() - ceilTasks;
-            log.debug("Tasks on worker {} is higher than ceiling, so revoking {} tasks", existing, numToRevoke);
-            for (int i = existing.tasksSize(); i > floorTasks && numToRevoke > 0; --i, --numToRevoke) {
+            int currentSize = forTask ? existing.tasksSize() : existing.connectorsSize();
+            int expectedSize;
+            if (existingWorkers.size() == 1 || currentSize == 1) expectedSize = ceil;
```
Comment (showuon) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java]: Why do we need this condition? And why under this case, we don't need to have `numberOfBiggers--`? After all, we assigned one ceiling to this worker, didn't we?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java [hunk 3]
```diff
@@ -562,50 +561,63 @@ private void resetDelay() {
...
-                resources.connectors().add(connectors.next());
-            }
-        }
+        return revoking;
+    }
 
+    static void computeRevoked(Map<String, ConnectorsAndTasks> revoking,
+                               Collection<WorkerLoad> existingWorkers,
+                               int numberOfTotalWorks,
+                               int numberOfActives,
```
Comment (showuon) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java]: Could we rename this parameters as `numberOfActiveTasks` to be more clear?

Comment (showuon) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java]: And they can be `final`

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java [hunk 4]
```diff
@@ -562,50 +561,63 @@ private void resetDelay() {
...
+                               boolean forTask) {
+        int floor = numberOfActives / numberOfTotalWorks;
+        int numberOfBiggers = numberOfActives % numberOfTotalWorks;
```
Comment (showuon) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java]: I believe you're trying to name it `numberOfRemainder`

**Review by showuon — COMMENTED** [review]
@chia7712 , thanks for the patch! This is a good finding! Left some comments. Thanks.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java (hunk 4 above)
Comment (showuon) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java]: Oh, I got what you mean, maybe `numberOfCeilingMembers` would be more clear?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java (hunk 2 above)
Comment (chia7712) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java]: That is a specify case as the node having single task is already in balance. We can move the job to another node but it does not balance anything but it brings extra down-time.

I will add comments for that case.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java (hunk 3 above)
Comment (chia7712) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java]: nice catch. will copy that.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java (hunk 4 above)
Comment (chia7712) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java]: ok. will copy that

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java [hunk 5]
```diff
@@ -562,50 +561,68 @@ private void resetDelay() {
...
-            Iterator<ConnectorTaskId> tasks = existing.tasks().iterator();
-            numToRevoke = existing.tasksSize() - ceilTasks;
-            log.debug("Tasks on worker {} is higher than ceiling, so revoking {} tasks", existing, numToRevoke);
-            for (int i = existing.tasksSize(); i > floorTasks && numToRevoke > 0; --i, --numToRevoke) {
+    static void computeRevoked(final Map<String, ConnectorsAndTasks> revoking,
+                               final Collection<WorkerLoad> existingWorkers,
+                               final int numberOfWorkers,
+                               final int numberOfActiveTasks,
+                               final boolean forTask) {
+        final int floor = numberOfActiveTasks / numberOfWorkers;
```
Comment (kkonstantine) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java]: Refactoring similar code to common methods is good and improves readability. 

But during code shuffling and especially if we add functional changes let's not apply also minor stylistic changes. 
We don't usually use `final` for local variables in Connect.

**Review by kkonstantine — COMMENTED** [review]
Thanks for the PR @chia7712 

I'd like to return for a detailed review after we exclude stylistic changes that are not necessary. Also, it'd be good to enhance the description besides the example and call out the corner cases for which logic has been added.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java (hunk 3 above)
Comment (kkonstantine) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java]: Let's not change local variables and arguments with `final` when there isn't this style in place. Sorry for the back and forth.

**Comment by chia7712** [comment]
@kkonstantine thanks for your feedback.

> I'd like to return for a detailed review after we exclude stylistic changes that are not necessary. Also, it'd be good to enhance the description besides the example and call out the corner cases for which logic has been added.

Sorry that I did not follow the code style in connect module. I have reverted all incorrect style. Also, I have updated the description.

**Review by kkonstantine — COMMENTED** [review]
Thanks for the quick turnaround @chia7712 !
Much easier to review line my line now (at least for me). 

I'll do my best to review by the end of the week.

**Pull request closed without merging**